sentinel/src/lib/purging.js

```diff
@@ -0,0 +1,445 @@
+              // for the contact or it's lineage via the `needs_signoff` path. Skip.
+              return;
+            }
+            const submitter = row.doc.contact && row.doc.contact._id;
+            if (!key && !subjectIds.includes(submitter)) {
+              // if the report doesn't have a subject, we want to process it when we hit the emit for the submitter.
+              // if the report submitter is not amongst our request keys, we hit an emit for the submitter's lineage. Skip.
+              return;
+            }
+          }
+
+          reportsByKey[key] = reportsByKey[key] || [];
+          reportsByKey[key].push(row.doc);
+        } else {
+          // messages only emit once, either their sender or receiver
+          key = row.key;
+          messagesByKey[key] = messagesByKey[key] || [];
+          messagesByKey[key].push(row.doc);
+        }
+
+        if (key && !subjectIds.includes(key)) {
+          // reports with errors that emitted the submitter id instead of the subject id
+          groups[key] = { contact: {}, subjectIds: [key], reports: [], messages: [] };
+        }
+        keys.add(key);
+      });
+
+      keys.forEach(key => {
+        if (!messagesByKey[key] && !reportsByKey[key]) {
+          return;
+        }
+        if (!key) {
+          // reports that have no subject are processed separately
+          reportsByKey[key].forEach(report => {
+            groups[report._id] = { contact: {}, reports: [report], messages: [] };
+          });
+          return;
+        }
+
+        const group = Object.values(groups).find(group => group.subjectIds.includes(key));
+        if (reportsByKey[key]) {
+          group.reports.push(...reportsByKey[key]);
+        }
+        if (messagesByKey[key]) {
+          group.messages.push(...messagesByKey[key]);
+        }
+      });
+
+      return getAlreadyPurgedDocs(rolesHashes, docIds);
+    })
+    .then(alreadyPurged => {
+      const toPurge = {};
+
+      Object.values(groups).forEach(group => {
+        rolesHashes.forEach(hash => {
+          toPurge[hash] = toPurge[hash] || {};
+
+          const allowedIds = [ group.contact._id, ...group.reports.map(r => r._id), ...group.messages.map(r => r._id)]
+            .filter(id => id);
+          if (!allowedIds.length) {
+            return;
+          }
+
+          const idsToPurge = purgeFn({ roles: roles[hash] }, group.contact, group.reports, group.messages);
+          if (!idsToPurge || !Array.isArray(idsToPurge) || !idsToPurge.length) {
+            return;
+          }
+
+          idsToPurge.forEach(id => {
+            if (!allowedIds.includes(id)) {
+              return;
+            }
+            toPurge[hash][id] = true;
+          });
+        });
+      });
+
+      return updatePurgedDocs(rolesHashes, docIds, alreadyPurged, toPurge);
+    })
+    .then(() => nextKey && batchedContactsPurge(roles, purgeFn, nextKey, nextKeyDocId));
+};
```

Comment:
This function is too long - is there any way to break it up into separate functions with descriptive names?

Reply:
I refactored it but I don't think I did a good job naming those auxiliary functions. Suggestions welcome!